### PR TITLE
plat-stm32mp1: add watchdog driver

### PR DESCRIPTION
**Pull request by etienne-lms**
This P-R implements stm32mp1 watchdog driver.
A later change will be posted once P-R and #5203 are merged, to implement watchdog timeout extension for this platform.

### CODE REVIEW

**Comment by etienne-lms** [comment]
forced pushed to fix CI: `.get_timeout` bad dependency.

File: core/drivers/stm32_iwdg.c [hunk 1]
```diff
@@ -0,0 +1,400 @@
...
+				   const void *fdt, int node)
+{
+	struct stm32_iwdg_otp_data otp_data = { };
+	TEE_Result res = TEE_SUCCESS;
+
+	res = stm32_iwdg_parse_fdt(iwdg, fdt, node);
+	if (res)
+		return res;
+
+	res = stm32_get_iwdg_otp_config(iwdg->base.pa, &otp_data);
```
Comment (jenswi-linaro) [inline core/drivers/stm32_iwdg.c:316]: This function is added in a later commit.

File: core/drivers/stm32_iwdg.c [hunk 2]
```diff
@@ -0,0 +1,400 @@
...
+ * @timeout - Watchdog elaspure timeout
+ * @wdt_chip - Wathcdog chip instance
+ * @link - Link in registered watchdog instance list
+ */
+struct stm32_iwdg_device {
+	struct io_pa_va base;
+	struct clk *clock;
+	struct clk *clk_lsi;
+	uint32_t flags;
+	unsigned long timeout;
```
Comment (jenswi-linaro) [inline core/drivers/stm32_iwdg.c:87]: Why is this a long?

File: core/drivers/stm32_iwdg.c [hunk 3]
```diff
@@ -0,0 +1,400 @@
...
+{
+	return iwdg->flags & IWDG_FLAGS_ENABLED;
+}
+
+/* Return counter value to related to input timeout in seconds, or 0 on error */
+static uint32_t iwdg_timeout_cnt(struct stm32_iwdg_device *iwdg,
+				 unsigned long to_sec)
+{
+	uint64_t reload = (uint64_t)to_sec * clk_get_rate(iwdg->clk_lsi);
+	uint32_t cnt = (reload / IWDG_PRESCALER_256) - 1;
```
Comment (jenswi-linaro) [inline core/drivers/stm32_iwdg.c:117]: Can this overflow?

File: core/drivers/stm32_iwdg.c (hunk 2 above)
Comment (etienne-lms) [reply core/drivers/stm32_iwdg.c:87]: this is driven by the timeout type used in _wdt.h_:
https://github.com/OP-TEE/optee_os/blob/5cbd8b3a709b1a2031f59ff10b8a60aefdea8d3b/core/include/drivers/wdt.h#L33-L35

File: core/drivers/stm32_iwdg.c (hunk 3 above)
Comment (etienne-lms) [reply core/drivers/stm32_iwdg.c:117]: ok, i'll make `cnt` a `uint64_t`

File: core/drivers/stm32_iwdg.c (hunk 1 above)
Comment (etienne-lms) [reply core/drivers/stm32_iwdg.c:316]: The driver expects the platform to provide this function, which is done by the next commit in the series.
A weak function in this source file is preferred?

File: core/drivers/stm32_iwdg.c (hunk 2 above)
Comment (jenswi-linaro) [reply core/drivers/stm32_iwdg.c:87]: It's an odd choice in the interface, but it makes sense to use the same type here.

File: core/drivers/stm32_iwdg.c [hunk 4]
```diff
@@ -114,13 +114,13 @@ static uint32_t iwdg_timeout_cnt(struct stm32_iwdg_device *iwdg,
...
 	uint64_t reload = (uint64_t)to_sec * clk_get_rate(iwdg->clk_lsi);
-	uint32_t cnt = (reload / IWDG_PRESCALER_256) - 1;
+	uint64_t cnt = (reload / IWDG_PRESCALER_256) - 1;
 
 	/* Be safe and expect any counter to be above 2 */
 	if (cnt > IWDG_CNT_MASK || cnt < 3)
 		return 0;
 
-	return cnt;
+	return (uint32_t)cnt;
```
Comment (jenswi-linaro) [inline core/drivers/stm32_iwdg.c:123]: I'm not sure this cast is needed.

**Review by jenswi-linaro — COMMENTED** [review]
`Acked-by: Jens Wiklander <jens.wiklander@linaro.org>`

**Pull request merged**